src/utils/recordedit-utils.ts

```diff
@@ -891,3 +884,106 @@ export function validateForeignkeyValue(
   }
 }
+
+export function disabledRowTooltip(disabledType: string): string {
+  let disabledTooltip = '';
+  if (disabledType === 'associated') {
+    disabledTooltip = MESSAGE_MAP.tooltip.otherRow;
+  } else if (disabledType === 'selected') {
+    disabledTooltip = MESSAGE_MAP.tooltip.otherInput;
+  }
+
+  return disabledTooltip;
+}
+
+/**
+ * Used to fetch the disabled tuples for a recordset modal picker used to associate rows of data
+ *
+ * @param prefillObject the prefill object in cookie storage fetched using queryParams
+ * @param domainRef the reference used in the modal picker that we want to disable rows for
+ * @param fkToLeaf the foreign key column from the association table that points to the leaf table
+ * @param fkToMain the foreign key column from the association table that points to the main table
+ * @param rowsUsedInForm
+ * @returns a function that returns a promise
+ */
+export function disabledTuplesPromise(prefillObject: PrefillObject, domainRef: any, fkToLeaf: any, fkToMain: any, rowsUsedInForm: SelectedRow[]) {
+  /**
+   * The existing rows in this p&b association must be disabled
+   * so users doesn't resubmit them.
+   */
+  return (
+    page: any,
+    pageLimit: number,
+    logStack: any,
+    logStackPath: string,
+    requestCauses?: any,
+    reloadStartTime?: any
+  ): Promise<{ page: any, disabledRows?: any }> => {
+    return new Promise((resolve, reject) => {
+      const disabledRows: any = [];
+
+      const disabledRowsFilters: any[] = [];
+      Object.keys(prefillObject.keys).forEach((key: string) => {
+        disabledRowsFilters.push({
+          source: [
+            { 'inbound': fkToLeaf.foreignKey.constraint_names[0] },
+            { 'outbound': fkToMain.foreignKey.constraint_names[0] },
+            fkToMain.foreignKey.mapping._to[0].name
+          ],
+          choices: [prefillObject.keys[key]]
+        });
+      });
+
+      let action = LogActions.LOAD,
+        newStack = logStack;
+      if (Array.isArray(requestCauses) && requestCauses.length > 0) {
+        action = LogActions.RELOAD;
+        newStack = LogService.addCausesToStack(logStack, requestCauses, reloadStartTime);
+      }
+
+      // using the service instead of the record one since this is called from the modal
+      const logObj = {
+        action: LogService.getActionString(action, logStackPath),
+        stack: newStack,
+      };
+
+      // fourth input: preserve the paging (read will remove the before if number of results is less than the limit)
+      domainRef
+        .addFacets(disabledRowsFilters)
+        .setSamePaging(page)
+        .read(pageLimit, logObj, false, true)
+        .then((newPage: any) => {
+          newPage.tuples.forEach((newTuple: any) => {
+            const index = page.tuples.findIndex((tuple: any) => {
+              return tuple.uniqueId === newTuple.uniqueId;
+            });
+
+            if (index > -1) {
+              page.tuples[index].disabledType = 'associated';
```

Comment:
We should treat the objects that we get from ermrestjs as immutable. Eventually ermrestjs will be converted to TypeScript, and we won't be able to do stuff like this.

I think you should do what we did for `selectedRows` prop for the `disabledRows`. So add a new `DisabledRow` type that has `tuple` and `disabledType` prop.